### PR TITLE
Client Credentials Support for Sharepoint

### LINKED ISSUE #524: [BC Idea]: Client Credentials authentication for SharePoint.
### BC Idea Link

https://experience.dynamics.com/ideas/idea/?ideaid=15430f25-7f70-ed11-a76e-0003ff45ac6d

### Description

I want to add support for Client Credentials Flow on the Sharepoint Client.

Related Issue: https://github.com/microsoft/ALAppExtensions/issues/22261

### I will provide the implementation for this BC Idea

- [X] I will provide the implementation for this BC Idea 

<br /><br />Internal work item: [AB#457003](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/457003)

### PR DESCRIPTION
**Pull request by pri-kise**
#### Summary
Add Client Credentials Support for Sharepoint

#### Work Item(s)
Fixes #524 

Fixes [AB#457003](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/457003)

Fixes [AB#522634](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/522634)

### CODE REVIEW

File: src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        if not AcquireToken(AccessToken, ErrorText) then
+            Error(ErrorText);
+        exit(AccessToken);
+    end;
+
+    local procedure AcquireToken(var AccessToken: SecretText; var ErrorText: Text): Boolean
+    var
+        OAuth2: Codeunit System.Security.Authentication.OAuth2;
+        FailedErr: Label 'Failed to retrieve an access token.';
+        //TODO: Check Authority Url
```
Comment (JesperSchulz) [inline src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al]: A todo in a non-draft PR? Shame! 😉😋

**Comment by JesperSchulz** [comment]
@darjoo / @WaelAbuSeada, would you mind looking at this PR with your security glasses put on? 😎

File: src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al (hunk 1 above)
Comment (pri-kise) [reply src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al]: I had problems setting up my new BC25 environment where I wanted to test this again.
I'll update this PR next week with some guideline on how to manually test this authentication method.

File: src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al [hunk 2]
```diff
@@ -82,4 +82,37 @@ codeunit 9142 "SharePoint Auth."
...
+    /// <summary>
+    /// Creates an authorization mechanism with the Client Credentials Grant Flow.
+    /// </summary>
+    /// <param name="AadTenantId">Azure Active Directory tenant ID</param>
+    /// <param name="ClientId">The Application (client) ID that the Azure portal - App registrations experience assigned to your app.</param>        
+    /// <param name="Certificate">The Base64-encoded certificate for the Application (client) configured in the Azure Portal - Certificates &amp; Secrets.</param>
+    /// <param name="CertificatePassword">Password for the certificate.</param>
+    /// <param name="Scope">A scope that you want the user to consent to.</param>
+    /// <returns>Codeunit instance implementing authorization interface</returns>
+    procedure CreateClientCredentials(AadTenantId: Text; ClientId: Text; Certificate: Text; CertificatePassword: SecretText; Scope: Text): Interface "SharePoint Authorization";
```
Comment (WaelAbuSeada) [inline src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]: use Secret text for the certificate too

File: src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al [hunk 3]
```diff
@@ -82,4 +82,37 @@ codeunit 9142 "SharePoint Auth."
...
+    /// <summary>
+    /// Creates an authorization mechanism with the Client Credentials Grant Flow.
+    /// </summary>
+    /// <param name="AadTenantId">Azure Active Directory tenant ID</param>
+    /// <param name="ClientId">The Application (client) ID that the Azure portal - App registrations experience assigned to your app.</param>        
+    /// <param name="Certificate">The Base64-encoded certificate for the Application (client) configured in the Azure Portal - Certificates &amp; Secrets.</param>
+    /// <param name="CertificatePassword">Password for the certificate.</param>
+    /// <param name="Scopes">A list of scopes that you want the user to consent to.</param>
+    /// <returns>Codeunit instance implementing authorization interface</returns>
+    procedure CreateClientCredentials(AadTenantId: Text; ClientId: Text; Certificate: Text; CertificatePassword: SecretText; Scopes: List of [Text]): Interface "SharePoint Authorization";
```
Comment (WaelAbuSeada) [inline src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]: use Secret text for the certificate too

**Review by WaelAbuSeada — DISMISSED** [review]

File: src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al (hunk 2 above)
Comment (pri-kise) [reply src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]: The certificate here should always be encrypted therefore this shouldn't have any security impact.
The procedures in the OAuth2 Codeunit doesn't use SecretText for the Certificate either.
@darjoo how should I proceed with this?

**Comment by pri-kise** [comment]
### How to test this new Autorization:

1. Create an App Registration in Azure with the following application permissions:
image
2. Create a suitable certifacte with password (see detailled guide below) + Upload the Certificate to your App Registration
3. Publish the System Application and the following sharepoint testing application (https://github.com/pri-kise/msdyn365bc-sharepointclient-test) in your docker environment
4. Open the Page Sharepoint Setup of the PTE App for manual testing
5. Upload the encrypted certifacte and set you password in the Setup Table
6. Test the connection with the suitable Actions


#### Detailled guide for creating a certifacte
_Originally posted by @jwikman in https://github.com/microsoft/ALAppExtensions/issues/22261#issuecomment-1917426607_

1. Create a KeyVault in the Azure Portal (or use an existing if you have one)
2. Generate a new certificate - KeyVault->Certificates->Generate/Import
  - Method: Generate
  - Name: Anything
  - Type: Self-signed
  - Subject: Any X.500 distinguished name
  - Content Type: `PKCS #12`
5. After it's been created, go to the certificate version and set `Enabled: Yes`
7. Then download the certificate as .pfx from the Certificate Version with the "Download in PFX/PEM format"

Now you've got a `PKCS #12` certificate saved as a .pfx file, without password (no good).

To get a password protected .pfx file I did this:
1. Start Microsoft Management Console by running `mmc.exe`
2. Add Snap-In `Certificates`, select "My user account" when prompted
3. Expand certificate tree to `Personal/Certificates`
4. Import your passwordless .pfx file via `Action/All Tasks/Import...`
8. Then right click on your certificate that should be in the `Personal/Certificates` store and select `All Tasks->Export...`
9. Include the Private Key
10. Set a password when asked
11. Save to file
12. Remove the certificate from `Personal/Certificates`

Now you've got a self-signed `PKCS #12` certificate saved as a .pfx file, with a password :)

File: src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al (hunk 1 above)
Comment (pri-kise) [reply src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al]: I updated my code an added instructions on how to test the new authorization.

**Comment by pri-kise** [comment]
@jwikman my PR is now ready to test.

File: src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al (hunk 2 above)
Comment (darjoo) [reply src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]: I'll update the OAuth2 module such that it takes in SecretText for the Certificate. Shouldn't take too long, I'll update your PR once I have that in as well.

**Comment by jwikman** [comment]
> @jwikman my PR is now ready to test.

Cool!
I'm choked with work that must be completed before the wave hits us, but a colleague of mine will try to test this out.

File: src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al (hunk 2 above)
Comment (JesperSchulz) [reply src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]: I think this one addresses this issue: #792 🙂

Comment (darjoo) [inline src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]:
```suggestion
    procedure CreateClientCredentials(AadTenantId: Text; ClientId: Text; Certificate: SecretText; CertificatePassword: SecretText; Scope: Text): Interface "SharePoint Authorization";
```

File: src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al (hunk 3 above)
Comment (darjoo) [inline src/System Application/App/SharePoint Authorization/src/SharePointAuth.Codeunit.al]:
```suggestion
    procedure CreateClientCredentials(AadTenantId: Text; ClientId: Text; Certificate: SecretText; CertificatePassword: SecretText; Scopes: List of [Text]): Interface "SharePoint Authorization";
```

File: src/System Application/App/SharePoint Authorization/src/SharePointAuthImpl.Codeunit.al [hunk 4]
```diff
@@ -11,12 +11,19 @@ codeunit 9143 "SharePoint Auth. - Impl."
...
-    [NonDebuggable]
     procedure CreateAuthorizationCode(EntraTenantId: Text; ClientId: Text; ClientSecret: SecretText; Scopes: List of [Text]): Interface "SharePoint Authorization";
     var
         SharePointAuthorizationCode: Codeunit "SharePoint Authorization Code";
     begin
         SharePointAuthorizationCode.SetParameters(EntraTenantId, ClientId, ClientSecret, Scopes);
         exit(SharePointAuthorizationCode);
     end;
+
+    procedure CreateClientCredentials(AadTenantId: Text; ClientId: Text; Certificate: Text; CertificatePassword: SecretText; Scopes: List of [Text]): Interface "SharePoint Authorization";
```
Comment (darjoo) [inline src/System Application/App/SharePoint Authorization/src/SharePointAuthImpl.Codeunit.al]:
```suggestion
    procedure CreateClientCredentials(AadTenantId: Text; ClientId: Text; Certificate: SecretText; CertificatePassword: SecretText; Scopes: List of [Text]): Interface "SharePoint Authorization";
```

File: src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+codeunit 9145 "SharePoint Client Credentials" implements "SharePoint Authorization"
+{
+    Access = Internal;
+    InherentEntitlements = X;
+    InherentPermissions = X;
+
+    var
+        ClientId: Text;
+        Certificate: Text;
```
Comment (darjoo) [inline src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al]:
```suggestion
        Certificate: SecretText;

```

File: src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        AadTenantId: Text;
+        Scopes: List of [Text];
+        CertificatePassword: SecretText;
+
+    procedure SetParameters(NewAadTenantId: Text; NewClientId: Text; NewCertificate: Text; NewCertificatePassword: SecretText; NewScopes: List of [Text])
```
Comment (darjoo) [inline src/System Application/App/SharePoint Authorization/src/SharePointClientCredentials.Codeunit.al]:
```suggestion
    procedure SetParameters(NewAadTenantId: Text; NewClientId: Text; NewCertificate: SecretText; NewCertificatePassword: SecretText; NewScopes: List of [Text])

```

**Review by darjoo — COMMENTED** [review]
I tried to update your branch directly, it was a little funky. So I'll just add the suggestions here and it should be simple enough to update.

**Comment by jwikman** [comment]
@pri-kise I think you would need to update your PR branch from microsoft:main, so we can access the new functions with SecretText parameters that Darrick added.

**Comment by darjoo** [comment]
@jwikman @pri-kise There is a test failure which I'm not sure is related. I updated his PR with the code suggestions above :)

[Build System Application Tests (Default) / System Application Tests (Default)](https://github.com/microsoft/BCApps/actions/runs/8373091623/job/22925897193#step:8:1704)
Function TestSendWithoutGetContent Assert.AreEqual failed. Expected:<502> (Integer). Actual:<200> (Integer). The response status code should be 200.

**Comment by pri-kise** [comment]
I update my branch from the main branch now.

**Comment by jwikman** [comment]
Then let's see if that also solved the test failure. I don't think that the failure is related to this PR...

**Review by darjoo — APPROVED** [review]

**Review by JesperSchulz — APPROVED** [review]

**Comment by pri-kise** [comment]
Is there any chance to get this backported to BC 24.x?
I orginally planned to get this addresse before the Deadline of BC24, but sadly I wasn't able to.

**Comment by JesperSchulz** [comment]
> Is there any chance to get this backported to BC 24.x? I orginally planned to get this addresse before the Deadline of BC24, but sadly I wasn't able to.

Absolutely. Consider this shipping with 24.1 in May!

**Comment by JesperSchulz** [comment]
PR created: #840.

**Comment by jwikman** [comment]
I was a few days late to this party... But now I finally got some time to test this out - and it works just as expected. 🥳

Thanks for doing this @pri-kise!

**Comment by JesperSchulz** [comment]
Since this is set to release with 24.1:
We'd need a feature details description with sample code. @pri-kise, could you provide a sentence or two incl. example - if possible?

**Pull request merged**